hack/lib/start.sh

```diff
@@ -0,0 +1,496 @@
+	                        --volume-dir="${VOLUME_DIR}"                            \
+	                        --etcd-dir="${ETCD_DATA_DIR}"                           \
+	                        --write-config="${SERVER_CONFIG_DIR}"                   \
+	                        --listen="${API_SCHEME}://${API_BIND_HOST}:${API_PORT}" \
+	                        --public-master="${API_SCHEME}://${PUBLIC_MASTER_HOST}:${API_PORT}"
+
+	os::start::internal::patch_master_config
+}
+readonly -f os::start::internal::configure_master
+
+function os::start::internal::patch_master_config() {
+	cp ${SERVER_CONFIG_DIR}/master/master-config.yaml ${SERVER_CONFIG_DIR}/master/master-config.orig.yaml
+	openshift ex config patch ${SERVER_CONFIG_DIR}/master/master-config.orig.yaml --patch="{\"etcdConfig\": {\"address\": \"${API_HOST}:${ETCD_PORT}\"}}" | \
+	openshift ex config patch - --patch="{\"etcdConfig\": {\"servingInfo\": {\"bindAddress\": \"${API_HOST}:${ETCD_PORT}\"}}}" | \
+	openshift ex config patch - --type json --patch="[{\"op\": \"replace\", \"path\": \"/etcdClientInfo/urls\", \"value\": [\"${API_SCHEME}://${API_HOST}:${ETCD_PORT}\"]}]" | \
+	openshift ex config patch - --patch="{\"etcdConfig\": {\"peerAddress\": \"${API_HOST}:${ETCD_PEER_PORT}\"}}" | \
+	openshift ex config patch - --patch="{\"etcdConfig\": {\"peerServingInfo\": {\"bindAddress\": \"${API_HOST}:${ETCD_PEER_PORT}\"}}}" | \
+	openshift ex config patch - --patch="{\"imagePolicyConfig\": {\"maxImagesBulkImportedPerRepository\": ${MAX_IMAGES_BULK_IMPORTED_PER_REPOSITORY:-5}}}" > ${SERVER_CONFIG_DIR}/master/master-config.yaml
+
+	# Make oc use ${MASTER_CONFIG_DIR}/admin.kubeconfig, and ignore anything in the running user's $HOME dir
+	export ADMIN_KUBECONFIG="${MASTER_CONFIG_DIR}/admin.kubeconfig"
+	CLUSTER_ADMIN_CONTEXT=$(oc config view --config=${ADMIN_KUBECONFIG} --flatten -o template --template='{{index . "current-context"}}'); export CLUSTER_ADMIN_CONTEXT
+	${USE_SUDO:+sudo} chmod -R a+rwX "${ADMIN_KUBECONFIG}"
+	echo "[INFO] To debug: export KUBECONFIG=$ADMIN_KUBECONFIG"
+}
+readonly -f os::start::internal::patch_master_config
+
+# os::start::server starts the OpenShift server, exports the PID of the OpenShift server and waits until openshift server endpoints are available
+# It is advised to use this function after a successful run of 'configure_os_server'
+#
+# Globals:
+#  - USE_SUDO
+#  - LOG_DIR
+#  - ARTIFACT_DIR
+#  - VOLUME_DIR
+#  - SERVER_CONFIG_DIR
+#  - USE_IMAGES
+#  - MASTER_ADDR
+#  - MASTER_CONFIG_DIR
+#  - NODE_CONFIG_DIR
+#  - API_SCHEME
+#  - API_HOST
+#  - API_PORT
+#  - KUBELET_SCHEME
+#  - KUBELET_HOST
+#  - KUBELET_PORT
+# Arguments:
+#  1 - API server version (ie. "v1.2.0")
+#  2 - Controllers version (ie. "v1.2.0")
+#  3 - Skip node start ("1" to skip node start)
+# Returns:
+#  - export OS_PID
+#  - export ETCD_PID
+#  - export API_SERVER_PID
+#  - export CONTROLLERS_PID
+#  - export NODE_PID
+function os::start::server() {
```

Comment:
I think the logic in this could be made a little clearer:

```sh
if [[ -n "${skip_node}" ]]; then
    if [[ -z "${api_server_version}" && -z "${controllers_version}" ]]; then
        # without a node or special version requirements
        os::start::master
    else
        # without a node but using special versions
        os::start::etcd
        os::start::api_server "${api_server_version}"
        os::start::controllers "${controllers_version}"
    fi
else
    if [[ -z "${api_server_version}" && -z "${controllers_version}" ]]; then
        # with a node but without special version requirements
        os::start::all_in_one
    else
        # with a node and using special versions
        os::start::etcd
        os::start::api_server "${api_server_version}"
        os::start::controllers "${controllers_version}"
        os::start::node
    fi
fi
```

Or maybe in the reverse is easier?
```sh
if [[ -z "${api_server_version}" && -z "${controllers_version}" ]]; then
    if [[ -z "${skip_node}" ]]; then
        os::start::all_in_one
    else
        os::start::master
    fi
else
    os::start::etcd
    os::start::api_server "${api_server_version}"
    os::start::controllers "${controllers_version}"
    if [[ -z "${skip_node}" ]]; then
        os::start::node
    fi
fi
```